### ISSUE TITLE
[python] Lockup when finalizing

### ISSUE THREAD

**Issue by tobwen** [protomaps/PMTiles#166]
## bug report
After creating a pmtiles file and adding 4 (or more) tiles to it, the Python code locks up. CPU usage is at 100% for hours, the process never finishes.

## code example
``` python
#!/usr/bin/env python3

from pmtiles.tile import zxy_to_tileid, tileid_to_zxy, TileType, Compression
from pmtiles.writer import Writer
import glob

with open("test.pmtiles", "wb") as f:
    writer = Writer(f)

    for filename in glob.iglob("tiles/5/**/*.png", recursive=True):
        base_filename, extension = filename.split(".")
        z, x, y = map(int, base_filename.split("/")[1:])

        tileid = zxy_to_tileid(z, x, y)
        try:
            tile = f"tiles/{z}/{x}/{y}.png"
            with open(tile, "rb") as img_file:
                writer.write_tile(tileid, img_file.read())
                print(f"added {tileid}, {tile}")
        except FileNotFoundError:
            pass

    print("finalizing...")
    writer.finalize(
        {
            "tile_type": TileType.PNG,
            "tile_compression": Compression.NONE,
            "min_zoom": 5,
            "max_zoom": 5,
            "min_lon_e7": int(5.86395260 * 1E7),
            "min_lat_e7": int(47.2904079 * 1E7),
            "max_lon_e7": int(15.0347900 * 1E7),
            "max_lat_e7": int(55.0547758 * 1E7),
            "center_zoom": 5,
            "center_lon_e7": int(11.58 * 1E7),
            "center_lat_e7": int(51.17 * 1E7)
        },
        {
            "attribution": 'n/a'
        },
    )
```

## output
I've interruped it after 2 hours.

```
added 1209, tiles/5/17/11.png
added 1212, tiles/5/17/10.png
added 1210, tiles/5/16/11.png
added 1211, tiles/5/16/10.png
finalizing...
# of addressed tiles: 4
# of tile entries (after RLE): 4
# of tile contents: 4
^CTraceback (most recent call last):
  File "test.py", line 39, in <module>
    "attribution": 'n/a'
  File "/usr/local/lib/python3.7/dist-packages/pmtiles-3.1.1-py3.7.egg/pmtiles/writer.py", line 87, in finalize
  File "/usr/local/lib/python3.7/dist-packages/pmtiles-3.1.1-py3.7.egg/pmtiles/writer.py", line 38, in optimize_directories
  File "/usr/local/lib/python3.7/dist-packages/pmtiles-3.1.1-py3.7.egg/pmtiles/tile.py", line 180, in serialize_directory
  File "/usr/local/lib/python3.7/dist-packages/pmtiles-3.1.1-py3.7.egg/pmtiles/tile.py", line 123, in write_varint
KeyboardInterrupt
```

## unegged version
```
^CTraceback (most recent call last):
  File "test.py", line 39, in <module>
    "attribution": 'n/a'
  File "/usr/local/lib/python3.7/dist-packages/pmtiles/writer.py", line 87, in finalize
    self.tile_entries, 16384 - 127
  File "/usr/local/lib/python3.7/dist-packages/pmtiles/writer.py", line 38, in optimize_directories
    test_bytes = serialize_directory(entries)
  File "/usr/local/lib/python3.7/dist-packages/pmtiles/tile.py", line 180, in serialize_directory
    write_varint(b_io, e.tile_id - last_id)
  File "/usr/local/lib/python3.7/dist-packages/pmtiles/tile.py", line 125, in write_varint
    b_io.write(bytes([towrite | 0x80]))
```

**Comment by tobwen** [comment]
## slashhammer method
``` python
def serialize_directory(entries):
    b_io = io.BytesIO()
    write_varint(b_io, len(entries))

    last_id = 0
    for e in entries:
        # dirty hack
        if (e.tile_id - last_id) < 0:
            break
        # dirty hack
        write_varint(b_io, e.tile_id - last_id)
        last_id = e.tile_id
```

works even on a larger dataset
```
finalizing...
# of addressed tiles: 37369
# of tile entries (after RLE): 37364
# of tile contents: 18514
Root dir bytes: 55
Leaves dir bytes: 74497
Num leaf dirs: 10
Total dir bytes: 74552
Average leaf dir bytes: 7449.7
Average bytes per addressed tile: 1.9950226123257244
```

**Comment by tobwen** [comment]
@bdon Sorry to bump this, but could you please take a look? I want to integrate PMTiles into some projects, but this problem is currently a blocker for me.

**Comment by bdon** [comment]
@tobwen right now a limitation of the python writer is you need to insert in increasing TileID order. In your example where you convert from a directory, first use `glob` to load all the filenames into memory, then convert each filename to TileID via `zxy_to_tileid`, then do the insert into the archive.

**Comment by bdon** [comment]
The other option is we allow un-clustered (sorted) archives for python, which will simply require sorting the entries in `Writer`. My approach so far has to been to always write in `clustered` order for C++, Go and Java, but depending on specific use case for Python users it might not be important.

**Comment by bdon** [comment]
Resolved in [python pmtiles v3.2.0](https://pypi.org/project/pmtiles/) which detects unordered writes, sets `clustered=false` and sorts entries when finalizing.
